### ISSUE TITLE
Weight age proportions (bubbles) by strata density

### ISSUE THREAD

**Issue by seananderson** [seananderson/gfplot#15]
Also consider weighting length proportions similarly assuming we switch to a binned length representation. If you still using a density representation of length, the density can be weighted.

See Holt et al. 2017 rock sole research document.

Could consider weighting the age-length, length-weight, and maturity fits by catch / density, but this hasn't been traditionally done and presumably would make little difference.

**Comment by seananderson** [comment]
Example: https://github.com/seananderson/pbs-synopsis/commit/71dbedd3f61a6f9ce34ec3cf4f25a60462d4002b

**Comment by seananderson** [comment]
Requires:
- [ ] bringing in commercial catch for each sample_id via fishing event
- [ ] bringing in survey strata and area (see `grouping_code` (`Fishing_Event` and `Grouping`) and `AREA_KM2` from `Grouping` and `GROUPING_DESC` from `Grouping` in GFBioSQL

**Comment by seananderson** [comment]
I integrated the SQL into the R code with https://github.com/seananderson/pbs-synopsis/commit/7cb199794d0918e0b674e4a8a25cb226f2024a24

**Comment by elisekeppel** [comment]
From talking with Maria it looks like it’s not simple to connect commercial trip/fishing event catch with the samples in GFBio – the trip id’s are different, sometimes the FOS trip id’s are in the comment field in GFBio, sometimes not, sometimes the Hail in or hail out number is in there, sometimes not, and they switched between linking with hail in and hail out numbers at some year. So… it’s not straightforward.

**Comment by seananderson** [comment]
I just added another column to retain in the survey samples query. We also need the survey area. Hopefully I have the column name right: https://github.com/seananderson/pbs-synopsis/commit/8395fe4857d1fc2862c8d204cf0828b3fbb33492

Regarding the commercial samples, hmm... There must be some way to do this because people have been weighting age proportions by catch in a few recent assessments. I'll send you the SQL queries I have from Rowan in an email.

**Comment by seananderson** [comment]
I wonder if there is a consistent way to do this for at least something like 1996 onwards?

**Comment by elisekeppel** [comment]
Shoot, yes there’s a column for area in the grouping codes table. I’ll run back and add that in before swinging by your place tonight. I just came home when I hit a wall and realized I need Norm or Maria to help me connect the gffos and gfbio databases.


On Dec 22, 2017, at 3:56 PM, Sean Anderson <notifications@github.com<mailto:notifications@github.com>> wrote:


I just added another column to retain in the survey samples query. We also need the survey area. Hopefully I have the column name right: 8395fe4<https://github.com/seananderson/pbs-synopsis/commit/8395fe4857d1fc2862c8d204cf0828b3fbb33492>

Regarding the commercial samples, hmm... There must be some way to do this because people have been weighting age proportions by catch in a few recent assessments. I'll send you the SQL queries I have from Rowan in an email.

**Comment by seananderson** [comment]
Almost done, by I think https://github.com/seananderson/pbs-synopsis/blob/master/inst/sql/get-survey-biology.sql needs

```
  ISNULL(USABILITY_CODE,1) IN (0, 1, 2, 6) AND
```

But `USABILITY_CODE` isn't available yet... need to come back to this.

**Comment by seananderson** [comment]
Got it, I think:
https://github.com/seananderson/pbs-synopsis/blob/master/R/weight-props.R
https://sean.updog.co/survey-sample-weighting.html

**Comment by seananderson** [comment]
Now doable with:

```r
# load data:

bio_specs <- readRDS("data-cache/all-commercial-bio.rds") %>% 
  filter(species_common_name %in% "pacific ocean perch")
catch <- readRDS("data-cache/all-catches.rds") %>% 
  filter(species_common_name %in% "pacific ocean perch")
survey_specimens <- readRDS("data-cache/all-survey-bio.rds") %>%
  filter(species_common_name %in% "pacific ocean perch") %>%
  filter(survey_series_id %in% 1)
survey_tows <- readRDS("data-cache/all-survey-spatial-tows.rds") %>%
  filter(species_common_name %in% "pacific ocean perch") %>%
  filter(survey_series_id %in% 1)

# weight proportions:
join_comps_commercial(bio_specs, catch, age) %>% 
  weight_comps()

join_comps_commercial(bio_specs, catch, length, bin_size = 2) %>% 
  weight_comps()

join_comps_survey(survey_specimens, survey_tows, age) %>% 
  weight_comps()

join_comps_survey(survey_specimens, survey_tows, length, bin_size = 2) %>% 
  weight_comps()
```

Adding a `group_by(species_common_name, survey_series_id)` would do all species and surveys.

**Comment by seananderson** [comment]
@andrew-edwards Could you look over the age/length frequency weighting code for me?

The meat of what needs to be examined is this:
https://github.com/seananderson/pbs-synopsis/blob/282b8062e8f01405ae8ea384565180bff12651c4/R/weight-props.R#L201-L252

and all other relevant code is in that file. The description for `weight_comps()` describes how the process works. If you load the package and look at the help for `weight_comps()` you'll seem some examples:

```r
install_github("seananderson/pbs-synopsis")
# or clone and:
# devtools::install() # full install locally, from base directory
library(PBSsynopsis)
# or:
# devtools::load_all() # quick install locally, from base directory

?weight_comps
```

```r
species <- "pacific ocean perch"
survey <- "Queen Charlotte Sound Synoptic Survey"

## Surveys:
survey_samples <- get_survsamples(species) %>%
  dplyr::filter(survey_series_desc == survey)
survey_tows <- get_survey(species) %>%
  dplyr::filter(survey_series_desc == survey)

surv_lengths <- tidy_comps_survey(survey_samples, survey_tows,
  value = length, bin_size = 2)
surv_lengths
weight_comps(surv_lengths)

surv_ages <- tidy_comps_survey(survey_samples, survey_tows, value = age)
surv_ages
weight_comps(surv_ages)

## Commercial:
com_samples <- get_commsamples(species)
com_catch <- get_catch(species)

com_lengths <- tidy_comps_commercial(com_samples, com_catch,
  value = length, bin_size = 2)
com_lengths
weight_comps(com_lengths)

com_ages <- tidy_comps_commercial(com_samples, com_catch,
  value = age)
com_ages
weight_comps(com_ages)

## These functions are pipe (%>%) friendly. E.g.:
tidy_comps_survey(survey_samples, survey_tows, value = age) %>%
  weight_comps()
```

Thanks, and feel free to bug me if you need more guidance. You'll need to be on the PBS network for the `get_*` functions to work.

Questions are:
1. Is this doing what is described in past Res Docs. e.g. p 161 of <http://waves-vagues.dfo-mpo.gc.ca/Library/363948.pdf>
2. Does this make sense or is there a different weighting scheme that would be better? It shouldn't be too hard to adapt within this framework now.

Although the code won't run with the current package version (based on old functions and data), there are some ideas for visualizing the output here: <https://sean.updog.co/survey-sample-weighting.html>

**Comment by andrew-edwards** [comment]
Sorry I won't have time for this until after hake meeting next week (I've just got back from a break and have to catch up and prepare for the meeting). Rowan normally does these calculations (I understand them when I read through the text), so I'll get him to have a look. He's away until 28th Feb. I don't think he has his own GitHub username but does use @pbs-software I think. I'll send him an email also.

**Comment by seananderson** [comment]
That’s fine, @andrew-edwards. Rowan @pbs-software the dplyr pipe syntax might look like Greek if you’re not familiar with it. Perhaps the most useful approach would be to have you, using your own code, generate weighted age frequencies from commercial and survey sources (one survey source) for an example stock and year range and we can compare. Assuming they don’t match exactly, it would be useful to compare the numerators and denominators in the weighting operations.

@andrew-edwards , when you come back to this, check in with me first. The package is evolving rapidly (including the repository and package changing names) so the above example code has changed.

**Comment by seananderson** [comment]
This file: https://github.com/seananderson/gfplot/blob/master/inst/test-weighting/test-weighting.Rmd
Rendered here: https://sean.updog.co/test-weighting.html

tests data provided by @haighr for redstripe rockfish against the implementation here. By some miracle, the weighted proportions seem to match.

The commercial comparison looks identical (see the 2nd example) with the exception of a small number of 29, 30, and 31 year-old redstripe in 2013 in Rowan's data that aren't in ours (see the blue dots that stand out). When we go back to the databases themselves, we can't find any redstripe rockfish of those ages in those years regardless of filtering criteria (actually I believe we found one that gets filtered out). The most likely scenario is that some data have been corrected in the database since Rowan pulled his version of the data.

The survey weighting looks slightly different in some cases, but when I inspect the bubbles that stand out the most, they are cases where we throw out many fish that came from survey tows that weren't considered 'useable' tows. We are using Norm's `proc_catmat_2011` procedure, which returns strata densities as used in the relative biomass indices, so we end up not retaining age and length samples in the composition weighting that came from tows excluded from the index calculation. So, I think we're just starting with slightly different raw data based on SQL filtering.

I'm reasonably convinced that the weighting code itself is correct since it matches almost perfectly for the commercial data and the same function is used to weight both the commercial and survey data. Furthermore, the calculation of the weighting numerators and denominators is pretty straightforward for the survey data.
